### DISCUSSION TITLE
External scenes vs. Hubs database

### DISCUSSION THREAD

**Discussion by Aaron-E-Walsh** [mozilla/hubs#3205] [General Questions & Discussion] [answered]
Hello, all. I wonder if anyone has insight into how external scenes (scenes hosted outside of hubs, such as on a standard web server) are handled by Hubs in terms of the Hubs database? 

For example, consider the following (imaginary) custom scene:

http://somewhereanywhereonline.com/scenes/myscene.glb

When I change my Hubs room scene using the "custom URL" option, and paste in the above URL, presumably the entire scene is imported automatically into the Hubs database, correct?

Now, imagine I change the room scene a few more times, such as to different scenes that I have created with Spoke (meaning they're not external to Hubs, but are internal -- already in the Hubs database). Finally I change the room scene once again to my my custom external scene, again using the URL above. 

Does this create a redundant copy of the scene in the Hubs database, because I have used the custom scene URL 2 different times in the course of changing room scenes?

I would imagine so, since in order to prevent database scene redundancy Hubs would likely need to hash the external scene each time it is used (in order to determine if the URL really does point to the same scene, in which case it can simply pull it out of the database, but if the scene is different in some way in which case a fresh import into the database is needed). And doing so would mean that it needs to download the entire external scene in order to obtain the hash, so it's not exactly efficient to use a hash. A quick-and-dirty way of doing this would be to compare URLs and also file sizes or HTTP file header info, but that's probably not accurate enough to be certain if the external scene is the same as one already in the database.

And so my guess is that Hubs is importing the scene each time the "custom URL" approach to loading a scene is used. And if this is the case each time I switch to a custom external scene using a URL (such as explained above) this would mean another copy of the scene is stored in the database, which can add up to a TON of redundant data over time (especially if the scenes are large, such as 10MB+ each).

Is there any way for me to check this in our Hubs Cloud database, or does anyone know how external scenes are handled when it comes to importing them into the database?

With thanks for your insights and advice,
Aaron

**Comment by robinkwilson** [comment]
@Aaron-E-Walsh **To clarify your question so I'm clear, is the below what you're asking? And please let me know if I'm missing something! :**

How are external scenes (hosted on an external url, not created in Spoke) handled by the Hubs db?

Example external custom scene: http://somewhereanywhereonline.com/scenes/myscene.glb

Steps:
1. In a Hubs room I change the scene via a "custom URL", using the above URL. _**-- Is the entire scene imported into the hubs db?**_
2. After, I change the scene to a few Spoke created scenes. -- already in the Hubs db.
3. Finally, I change the room scene back to the above URL.

**_Is the scene at the URL copied into the Hubs db twice or is it cached?_**

**Comment by Aaron-E-Walsh** [comment]
Hello, Robin. Thank you for asking, and for answering my question. Yes, as
you have summarized it that is indeed my question (simply put, does Hubs
copy an external scene into the db more than once or is it cached upon
first usage and pulled from the db every time it is referenced via external
URL from that point forward).

With thanks again,
Aaron

**Comment by brianpeiris** [comment, marked as answer]
Hi Aaron,

Hubs does not import scenes into its database when you use a custom scene URL. We simply store the URL itself in our database, and use a proxy to make that URL accessible. The proxy respects the caching rules and behavior determined by the original server, and that cache is typically on the original server or the user's browser cache, so Hubs is not involved in caching the custom scene either.

As such, you don't have to worry about switching to/from custom scene URLs and you can use them as many times as you like.

Hope that answers your question.

**Comment by Aaron-E-Walsh** [comment]
Ah, I see. This approach makes plenty of sense, and as it doesn't store any information other than the external scene URL it's very lightweight and so I no longer have concerns about the database being taxed. 

With thanks to you both for clarifying how hubs handles external scenes vis-à-vis the database.

Regards,
Aaron
